lib/cartopy/util.py

```diff
@@ -99,3 +101,298 @@ def add_cyclic_point(data, coord=None, axis=-1):
+
+    Examples
+    --------
+    Adding a cyclic point to a data array, where the cyclic dimension is
+    the right-most dimension.
+    .. testsetup::
+        >>> from distutils.version import LooseVersion
+        >>> import numpy as np
+        >>> if LooseVersion(np.__version__) >= '1.14.0':
+        ...     # To provide consistent doctests.
+        ...     np.set_printoptions(legacy='1.13')
+    >>> import numpy as np
+    >>> data = np.ones([5, 6]) * np.arange(6)
+    >>> cyclic_data = add_cyclic(data)
+    >>> print(cyclic_data)  # doctest: +NORMALIZE_WHITESPACE
+    [[ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]]
+
+    Adding a cyclic point to a data array and an associated coordinate.
+    >>> lons = np.arange(0, 360, 60)
+    >>> cyclic_data, cyclic_lons = add_cyclic(data, coord=lons)
+    >>> print(cyclic_data)  # doctest: +NORMALIZE_WHITESPACE
+    [[ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]]
+    >>> print(cyclic_lons)
+    [  0  60 120 180 240 300 360]
+
+    Adding a cyclic point to a data array and an associated 2-dimensional
+    coordinate.
+    >>> lons = np.arange(0, 360, 60)
+    >>> lats = np.arange(-90, 90, 180/5)
+    >>> lon2d, lat2d = np.meshgrid(lons, lats)
+    >>> cyclic_data, cyclic_lon2d = add_cyclic(data, coord=lon2d)
+    >>> print(cyclic_data)  # doctest: +NORMALIZE_WHITESPACE
+    [[ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]]
+    >>> print(cyclic_lon2d)
+    [[  0  60 120 180 240 300 360]
+     [  0  60 120 180 240 300 360]
+     [  0  60 120 180 240 300 360]
+     [  0  60 120 180 240 300 360]
+     [  0  60 120 180 240 300 360]]
+
+    Adding a cyclic point to a data array and an associated 2-dimensional
+    coordinate and a second raw variable.
+    >>> lons = np.arange(0, 360, 60)
+    >>> lats = np.arange(-90, 90, 180/5)
+    >>> lon2d, lat2d = np.meshgrid(lons, lats)
+    >>> cyclic_data, cyclic_lon2d, cyclic_lat2d = add_cyclic(
+    ...     data, coord=lon2d, rowcoord=lat2d)
+    >>> print(cyclic_data)  # doctest: +NORMALIZE_WHITESPACE
+    [[ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]
+     [ 0. 1. 2. 3. 4. 5. 0.]]
+    >>> print(cyclic_lon2d)
+    [[  0  60 120 180 240 300 360]
+     [  0  60 120 180 240 300 360]
+     [  0  60 120 180 240 300 360]
+     [  0  60 120 180 240 300 360]
+     [  0  60 120 180 240 300 360]]
+    >>> print(cyclic_lat2d)
+    [[-90. -90. -90. -90. -90. -90. -90.]
+     [-54. -54. -54. -54. -54. -54. -54.]
+     [-18. -18. -18. -18. -18. -18. -18.]
+     [ 18.  18.  18.  18.  18.  18.  18.]
+     [ 54.  54.  54.  54.  54.  54.  54.]]
+
+    Not adding a cyclic point if cyclic point detected in coord.
```

Comment:
Do we want to warn here? I could see both sides, one where someone wants to call `add_cyclic` every time and they would be annoyed with a warning, but also someone else may want to call `add_cyclic` on their data and expect it to be expanded when it wasn't and now the returns aren't the shape they were expecting.

Reply:
I do not want to warn here. I wrote the code to use it in one of my projects, of course [ncvue](https://github.com/mcuntz/ncvue). And there I am your case 1, i.e. I call it every the time. I regard the second case as erroneous anyway.
I could  imagine a compromise by making _has_cyclic a public function (calling it has_cyclic).